deps/v8/include/v8.h

```diff
@@ -5381,6 +5381,17 @@ class V8_EXPORT Isolate {
   static Isolate* GetCurrent();
 
   /**
+   * Custom callback used by embedders to help V8 determine if it should abort
+   * when it throws and no internal handler can catch the exception.
+   * If FLAG_abort_on_uncaught_exception is true, then V8 will abort if either:
+   * - no custom callback is set.
+   * - the custom callback set returns true.
+   * Otherwise it won't abort.
+   */
+  typedef bool (*abort_on_uncaught_exception_t)(Isolate*);
```

Comment:
Before I continue commenting, I just want to say that I'm thinking out loud and trying to find the right set of abstractions. I think your suggestions are great and help move this PR forward. I really really appreciate them and don't want to sound like I'm dismissing them, and I would appreciate if we could continue this constructive conversation.

The problem with aborting in node or in V8 when using `--abort-on-uncaught-exception` is consistency. For instance, to get the same exit codes and signals in both cases would require to use the same abort implementation. In this case, it would mean using  the same implementation as `base::OS::Abort()`, which we can't use since it's not part of V8's public API (unless I'm missing something). This implementation has changed in the past and we would need to update our implementation to match V8's implementation if it changes again.

While it might not seem like a big issue, to me it indicates that we haven't found the right abstraction.

Now that I've implemented the change you suggested to have an enum rather than a boolean as the return value type for `UncaughtExceptionCallback`, I have the same concerns. When `FLAG_abort_on_uncaught_exception` is set, it doesn't make sense to me to do anything than aborting or not aborting (as an exception to the rule). Encoding the different behaviors at that point in an enum means that the `UncaughtExceptionCallback` in node is written like following:
```
Isolate::UncaughtExceptionPolicy OnUncaughtException(Isolate* isolate) {
  if (need_to_not_abort) {
   return Isolate::UncaughtExceptionPolicy::NO_ABORT;
  } else {
   return Isolate::UncaughtExceptionPolicy::ABORT;
  }
}
```
My concern with this implementation is that it's written like if node could do something else than aborting or not aborting, like swallowing the exception. But in reality V8 will abort or exit, but nothing else would make sense.

> In a perfect world --abort_on_uncaught_exception wouldn't exist in V8 and just be handled by the embedder.

I don't know. node is maybe the only embedder taking advantage of this command line switch today, but to me it seems general enough to be used by any other embedder. For instance, even though I haven't done it yet, using d8 with `--abort-on-uncaught-exception` would be a good way for me to test mdb_v8 with recent V8 versions that haven't been integrated into node.

mdb_v8 is also a post-mortem debugger for any V8 embedder, not just node. These examples still revolve around mdb_v8, but I think it illustrates that `--abort-on-uncaught-exception` is a sound concept for any V8 embedder. 

So I'm back to thinking that having the `UncaughtExceptionCallback` return a `bool` is better than having it return an `enum`, but as I said at the beginning of this comment I'm open to any further feedback. It might also help to continue the conversation on IRC for a bit, I'm "jgi" in #io.js, #libuv and #v8.